gh-139165: Make Py_SIZE, Py_IS_TYPE,Py_ SET_SIZE regular functions in stable ABI

encukou opened the pull request:
Group them together with Py_TYPE & Py_SET_TYPE to cut down on repetitive preprocessor macros.
Format repetitive definitions in object.c more concisely.

Py_SET_TYPE is still left out of the Limited API.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-139165 -->
* Issue: gh-139165
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--139166.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

encukou committed:
Make Py_{SIZE,IS_TYPE,SET_SIZE} regular functions in stable ABI

Group them together with Py_TYPE & Py_SET_TYPE to cut down
on repetitive preprocessor macros.
Format repetitive definitions in object.c more concisely.

Py_SET_TYPE is still left out of the Limited API.

vstinner commented on Include/object.h:
```diff
@@ -323,9 +308,31 @@ static inline void Py_SET_SIZE(PyVarObject *ob, Py_ssize_t size) {
+#  define Py_SET_SIZE(ob, size) _Py_SET_SIZE_impl(_PyVarObject_CAST(ob), (size))
+#  define Py_SET_TYPE(ob, type) Py_SET_TYPE(_PyObject_CAST(ob), type)
+#elif Py_LIMITED_API+0 < _Py_PACK_VERSION(3, 15)
+#  if Py_LIMITED_API+0 < _Py_PACK_VERSION(3, 14)
```
Do you mean `>=`?

encukou replied:
No, but I can expand the comments :)

vstinner commented on Objects/object.c:
```diff
@@ -3405,3 +3387,16 @@ _PyObject_VisitType(PyObject *op, visitproc visit, void *arg)
+#undef Py_SIZE
+#undef Py_IS_TYPE
+#undef Py_SET_SIZE
+PyTypeObject* Py_TYPE(PyObject *ob) { return _Py_TYPE_impl(ob); }
+Py_ssize_t Py_REFCNT(PyObject *ob) { return _Py_REFCNT(ob); }
+Py_ssize_t Py_SIZE(PyObject *o) { return _Py_SIZE_impl(o); }
+int Py_IS_TYPE(PyObject *o, PyTypeObject *t) { return _Py_IS_TYPE_impl(o, t); }
+void Py_SET_SIZE(PyVarObject *o, Py_ssize_t s) { _Py_SET_SIZE_impl(o, s); }
```
Why not using the regular PEP 7 formatting?

```suggestion
PyTypeObject*
Py_TYPE(PyObject *ob)
{
    return _Py_TYPE_impl(ob);
}

Py_ssize_t
Py_REFCNT(PyObject *ob)
{
    return _Py_REFCNT(ob);
}

Py_ssize_t
Py_SIZE(PyObject *o)
{
    return _Py_SIZE_impl(o);
}

int
Py_IS_TYPE(PyObject *o, PyTypeObject *t)
{
    return _Py_IS_TYPE_impl(o, t);
}

void
Py_SET_SIZE(PyVarObject *o, Py_ssize_t s)
{
    _Py_SET_SIZE_impl(o, s);
}
```

encukou replied:
I think it's easier to read in the compact form. If every line has the same structure it's easier to scan that the names match.
Plus, the comment is more visible as you don't need to scroll up to see it.

vstinner replied:
Ok, as you want. I don't have a strong preference.

encukou committed:
Adjust comments

encukou committed:
Merge in the main branch

encukou committed:
Add blurb

encukou committed:
Fix reference

encukou committed:
Merge in the main branch

encukou committed:
Update for _Py_OPAQUE_PYOBJECT

- Don't expose the "real" functions for old Limited API
- Move implementations below defines so that they pick up
  the right Py_TYPE definition

encukou committed:
Merge in the main branch

encukou committed:
Note that PyObject fields are part of stable ABI

vstinner reviewed (approved):
LGTM

davidhewitt commented on Include/object.h:
```diff
@@ -265,56 +265,72 @@ _Py_IsOwnedByCurrentThread(PyObject *ob)
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= _Py_PACK_VERSION(3, 14)
 PyAPI_FUNC(PyTypeObject*) Py_TYPE(PyObject *ob);
+#endif
+#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 >= _Py_PACK_VERSION(3, 15)
+PyAPI_FUNC(Py_ssize_t) Py_SIZE(PyObject *ob);
+PyAPI_FUNC(int) Py_IS_TYPE(PyObject *ob, PyTypeObject *type);
+PyAPI_FUNC(void) Py_SET_SIZE(PyVarObject *ob, Py_ssize_t size);
+#endif
```
I'm just checking PyO3's FFI definitions against 3.15b1 and it seems to me that this PR has changed these symbols to _always_ be exported functions on 3.15.

That seems to differ with the issue description, which apparently intended to keep the static inline functions for the non-limited API?

vstinner replied:
Extract of object.c in the 3.15 branch:

```c
static inline int
_Py_IS_TYPE_impl(PyObject *ob, PyTypeObject *type)
{
    return Py_TYPE(ob) == type;
}

#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 < _Py_PACK_VERSION(3, 11)
#  define Py_IS_TYPE(ob, type) _Py_IS_TYPE_impl(_PyObject_CAST(ob), (type))
#elif Py_LIMITED_API+0 < _Py_PACK_VERSION(3, 15)
#  define Py_IS_TYPE(ob, type) _Py_IS_TYPE_impl((ob), (type))
#else
// Limited API 3.15+: use function calls
#endif
```

If Py_LIMITED_API is not defined or if it's defined to Python 3.14 or older, Py_IS_TYPE() is implemented as a static inline function. The code has been modified since this PR.

Would you mind to elaborate which API is implemented as a function call? Using the limited C API or not? Which limited C API version?

davidhewitt replied:
You're right, sorry for the confusion. I think the tooling I was using to check had a bug, and then I saw this diff and got myself confused. 🤦 

Even looking at this diff again, the `#define` macros just below starting on 292 will implement as a static inline function.

Sorry for the noise!

vstinner replied:
No problem. Thanks for double checking.
